# Analysis of TARGET-seq+ single cell genotyping data

# For calling single cell genotypes at a single mutation locus, integrating data from gDNA and mRNA/cDNA amplicons.

# Sample NOC131 DNMT3A_pI780T

#### Load the relevant libraries ####
library(tidyverse)
library(cowplot)
theme_set(theme_cowplot())

## Set key variables ####
# This makes it easier to use code for other files

# Paths to genotyping allelic counts files generated by the TARGET-seq single cell genotyping pipeline (SCpipeline)
# These can be found in the "Summarize" directory
gDNA_filepath <- "data/genotyping/sample_NOC131/gDNA_DNMT3A_pI780T.txt"
mRNA_filepath <- "data/genotyping/sample_NOC131/mRNA_DNMT3A_pI780T.txt"


# Path to metadata file
metadata_filepath <- "data/genotyping/genotyping_metadata.txt"

# Name of control sample
control <- "NOC153"

# Test sample being analysed
sample <- "NOC131"

# Mutation base change
# This sets which columns in the file to use as reference and variant alleles
ref <- quo(A)
mut <- quo(G)


## READ IN FILES #####

# Read in the csv file for each sequencing run:
genotyping_gDNA <- read_tsv(gDNA_filepath)
genotyping_mRNA <- read_tsv(mRNA_filepath)

# Read in the metadata file
metadata <- read_tsv(metadata_filepath)


# Separate the id column to get the cell ID and tidy up
genotyping_gDNA <- genotyping_gDNA %>% 
  separate(id, c("id", "mutation"), sep = "[.]") %>% 
  separate(id, c("Cell", NA), sep = "[_]") %>% 
  rename(chromosome = bp, position = ref)
str(genotyping_gDNA)

genotyping_mRNA<- genotyping_mRNA %>% 
  separate(id, c("id", "mutation"), sep = "[.]") %>% 
  separate(id, c("Cell", NA), sep = "[_]") %>% 
  rename(chromosome = bp, position = ref)
str(genotyping_mRNA)


# Mutation name
mutation_name <- genotyping_gDNA$mutation[1]



### JOIN METADATA TO VARIANT CALLS ####

# Join the library metadata to the genotyping data]
genotyping_gDNA <- genotyping_gDNA %>% 
  full_join(metadata, by = "Cell")

genotyping_mRNA <- genotyping_mRNA %>% 
  full_join(metadata, by = "Cell") 


# List of plates for the relevant sample
sample_plates <- genotyping_gDNA %>% 
  filter(Sample == sample) %>% 
  pull(Plate) %>% 
  unique()

# Keep only the plates relevant to this mutation
genotyping_gDNA <- genotyping_gDNA %>% 
  filter(Plate %in% sample_plates)
genotyping_mRNA <- genotyping_mRNA %>% 
  filter(Plate %in% sample_plates)


# Where no reads were detected and no fastq file exists, the sample will not be in the genotyping file
# Fill in the read counts with zeros for these, to allow them to be included, with no coverage
genotyping_gDNA <- genotyping_gDNA %>% 
  replace_na(list(A = 0, G = 0, C = 0, T = 0, del = 0, ins = 0))
genotyping_mRNA <- genotyping_mRNA %>% 
  replace_na(list(A = 0, G = 0, C = 0, T = 0, del = 0, ins = 0))

# Calculate the coverage and mutant VAF 
genotyping_gDNA <- genotyping_gDNA %>% 
  mutate(coverage = A + G + C + T + del + ins) %>% 
  mutate(VAF = !!mut / coverage)

genotyping_mRNA <- genotyping_mRNA %>% 
  mutate(coverage = A + G + C + T + del + ins) %>% 
  mutate(VAF = !!mut / coverage)



## FILTERING #####


# Calculate the coverage in blank wells to set the threshold for detection
genotyping_gDNA_blanks <- genotyping_gDNA %>% 
  filter(Sample == "Blank")

gDNA_blank_cov <- genotyping_gDNA_blanks %>% 
  group_by(Plate) %>% 
  summarise(mean_coverage = mean(coverage),
            max_coverage = max(coverage),
            min_coverage = min(coverage),
  ) %>% arrange(desc(max_coverage))
gDNA_blank_cov

genotyping_mRNA_blanks <- genotyping_mRNA %>% 
  filter(Sample == "Blank")

mRNA_blank_cov <- genotyping_mRNA_blanks %>% 
  group_by(Plate) %>% 
  summarise(mean_coverage = mean(coverage),
            max_coverage = max(coverage),
            min_coverage = min(coverage),
  ) %>% 
  arrange(desc(max_coverage))
mRNA_blank_cov

# Take a look at the coverage for non-blank samples
genotyping_gDNA %>% 
  filter(Sample != "Blank") %>% 
  select(Cell, A, G, C, T, coverage, VAF) %>% 
  arrange(desc(coverage))

genotyping_mRNA %>% 
  filter(Sample != "Blank") %>%   
  select(Cell, A, G, C, T, coverage, VAF) %>% 
  arrange(desc(coverage))

# Plot the coverage
# gDNA
ggplot(genotyping_gDNA %>% mutate(Sample=if_else(Sample == "NOC153", "Control", Sample)),
       aes(x= Plate, y=(coverage+1), fill = Sample, colour = Sample)) +
  geom_violin(scale = "width", alpha = 0.4) +
  geom_jitter(position = position_jitterdodge(jitter.width = 0.1, jitter.height = 0, dodge.width = 0.9), alpha = 0.2)+
  labs(title = paste(mutation_name, "gDNA amplicon"), y = "Coverage")+
  scale_y_log10()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# mRNA
ggplot(genotyping_mRNA %>% mutate(Sample=if_else(Sample == "NOC153", "Control", Sample)), 
       aes(x= Plate, y=(coverage+1), fill = Sample, colour = Sample)) +
  geom_violin(scale = "width", alpha = 0.4) +
  geom_jitter(position = position_jitterdodge(jitter.width = 0.1, jitter.height = 0, dodge.width = 0.9), alpha = 0.2)+
  labs(title = paste(mutation_name, "mRNA amplicon"), y = "Coverage")+
  scale_y_log10()+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


##### Filter the dataset to remove failed libraries ####

# We set the minimum coverage for gDNA and mRNA amplicons
gDNA_min_coverage <- 50
mRNA_min_coverage <- 40

# Filter out cells with coverage below the threshold
genotyping_gDNA_filtered <- genotyping_gDNA %>% 
  filter(coverage >= gDNA_min_coverage)
genotyping_mRNA_filtered <- genotyping_mRNA %>% 
  filter(coverage >= mRNA_min_coverage)



##### Check how many cells pass QC ####

qc_results <- genotyping_gDNA %>% 
  group_by(Sample) %>% 
  summarise(Total = n(),
            Passed = sum(coverage >= gDNA_min_coverage),
            Failed = sum(coverage < gDNA_min_coverage),
            Pass_rate = Passed/Total,
            Mean_coverage = mean(coverage)) %>% 
  mutate(Amplicon = "gDNA")

qc_results <- genotyping_mRNA %>% 
  group_by(Sample) %>% 
  summarise(Total = n(),
            Passed = sum(coverage >= mRNA_min_coverage),
            Failed = sum(coverage < mRNA_min_coverage),
            Pass_rate = Passed/Total,
            Mean_coverage = mean(coverage)) %>% 
  mutate(Amplicon = "mRNA") %>% 
  bind_rows(qc_results) %>% 
  filter(Sample != "Blank") %>% 
  mutate(Mutation = mutation_name)
qc_results

ggplot(qc_results, aes(Amplicon, Pass_rate, fill = Sample))+
  geom_col(position = "dodge")+
  labs(title = mutation_name, y = "Cells passing QC %")+
  scale_fill_brewer()



## CALLING MUTANT CELLS #####

# Plot mutant VAF

ggplot(genotyping_gDNA_filtered, aes(x = Sample, y = VAF, fill = Sample)) +
  geom_point(position = "jitter")+
  labs(title = paste(mutation_name, "gDNA amplicon"), y = "VAF")

ggplot(genotyping_mRNA_filtered, aes(x = Sample, y = VAF, fill = Sample)) +
  geom_point(position = "jitter") +
  labs(title = paste(mutation_name, "mRNA amplicon"), y = "VAF")


# To define the appropriate VAF thresholds for mutation calling, we use the VAF distribution from WT control cells to determine the noise at each locus.

# We define two thresholds for calling mutations:
# 1. The lower VAF threshold for calling a cell WT is defined as: mean(VAF) + 3 * SD(VAF)
# 2. The upper VAF threshold for calling a cell mutant is defined as: mean(VAF) + 3 * SD(VAF) + 0.01
# Cells between these thresholds are called "Undetermined"

# Calculate the mean and SD of the WT control VAF distribution and use these to calculate the VAF thresholds

gDNA_control_VAF <- genotyping_gDNA_filtered %>% 
  filter(Sample == control) %>% 
  summarise(mean_VAF = mean(VAF), SD_VAF = sd(VAF),
            max_VAF = max(VAF),
            min_VAF = min(VAF),
            WT_threshold = mean_VAF + 3*SD_VAF,
            MUT_threshold = WT_threshold + 0.01)
gDNA_control_VAF

mRNA_control_VAF <- genotyping_mRNA_filtered %>% 
  filter(Sample == control) %>% 
  summarise(mean_VAF = mean(VAF), SD_VAF = sd(VAF),
            max_VAF = max(VAF),
            min_VAF = min(VAF),
            WT_threshold = mean_VAF + 3*SD_VAF,
            MUT_threshold = WT_threshold + 0.01)
mRNA_control_VAF


# Furthermore, we required a minimum number of 10 mutant reads for a cell to be called mutant. 
gDNA_min_mut_reads <- 10
mRNA_min_mut_reads <- 10


# Plot these thresholds on the VAF distribution
ggplot(genotyping_gDNA_filtered, aes(x = fct_recode(Sample, "Control" ="NOC153"), y = VAF)) +
  geom_jitter(height = 0, size = 0.75)+
  geom_hline(aes(yintercept = gDNA_control_VAF$WT_threshold), linetype = "longdash", color = "#E69F00") +
  geom_hline(aes(yintercept = gDNA_control_VAF$MUT_threshold), linetype = "longdash", color = '#56B4E9') +
  theme(plot.title = element_text(hjust = 0.5))+
  labs(title = paste(str_replace(mutation_name, "_", " "), "gDNA"), y = "VAF", x = 'Sample')

ggplot(genotyping_mRNA_filtered, aes(x = fct_recode(Sample, "Control" ="NOC153"), y = VAF)) +
  geom_jitter(height = 0, size = 0.75)+
  geom_hline(aes(yintercept = mRNA_control_VAF$WT_threshold), linetype = "longdash", color = "#E69F00") +
  geom_hline(aes(yintercept = mRNA_control_VAF$MUT_threshold), linetype = "longdash", color = '#56B4E9') +
  theme(plot.title = element_text(hjust = 0.5))+
  labs(title = paste(str_replace(mutation_name, "_", " "), "mRNA"), y = "VAF", x = 'Sample')


##### Assign genotypes based on VAF thresholds ####

# Call genotypes in gDNA amplicons
genotyping_gDNA_filtered <- genotyping_gDNA_filtered %>% 
  mutate(genotype = case_when(
    # If VAF is below WT threshold, cell is WT
    VAF < gDNA_control_VAF$WT_threshold ~ "WT",  
    # If VAF is above MUT threshold and has ≥ 10 MUT reads, cell is MUT
    VAF >= gDNA_control_VAF$MUT_threshold & !!mut >= gDNA_min_mut_reads ~ "Mut", 
    # If VAF is above MUT threshold but has < 10 MUT reads, cell is Undetermined
    VAF >= gDNA_control_VAF$MUT_threshold & !!mut < gDNA_min_mut_reads  ~ "Undetermined", 
    # If VAF is between MUT and WT thresholds, it is Undetermined
    VAF < gDNA_control_VAF$MUT_threshold & VAF >= gDNA_control_VAF$WT_threshold ~ "Undetermined")) 

# Call genotypes in mRNA amplicons
genotyping_mRNA_filtered <- genotyping_mRNA_filtered %>% 
  mutate(genotype = case_when(
    # If VAF is below WT threshold, cell is WT
    VAF < mRNA_control_VAF$WT_threshold ~ "WT",  
    # If VAF is above MUT threshold and has ≥ 10 MUT reads, cell is MUT
    VAF >= mRNA_control_VAF$MUT_threshold & !!mut >= mRNA_min_mut_reads ~ "Mut", 
    # If VAF is above MUT threshold but has < 10 MUT reads, cell is Undetermined
    VAF >= mRNA_control_VAF$MUT_threshold & !!mut < mRNA_min_mut_reads  ~ "Undetermined", 
    # If VAF is between MUT and WT thresholds, it is Undetermined
    VAF < mRNA_control_VAF$MUT_threshold & VAF >= mRNA_control_VAF$WT_threshold ~ "Undetermined")) 


# Plot thresholds on a log scale of VAF and mutant read count to check the distribution of test and control cells
ggplot(genotyping_gDNA_filtered, aes(x = !!mut+0.1, y = VAF+0.0001, colour = Sample)) +
  geom_rect(aes(xmin = gDNA_min_mut_reads, xmax =Inf,  ymin = gDNA_control_VAF$MUT_threshold, ymax = Inf),
            fill = "#f0f0f0", colour = NA)+geom_point(position = "jitter")+
  geom_vline(aes(xintercept = gDNA_min_mut_reads),  linetype = "longdash")+
  geom_hline(aes(yintercept = gDNA_control_VAF$MUT_threshold),  linetype = "longdash") +
  geom_hline(aes(yintercept = gDNA_control_VAF$WT_threshold),  linetype = "longdash") +
  labs(title = paste(mutation_name, "gDNA threshold"), y = "VAF", x = "Mutant reads")+
  scale_x_log10()+scale_y_log10()
ggplot(genotyping_gDNA_filtered %>% filter(Sample !=control), 
       aes(x = !!mut+0.1, y = VAF+0.0001, colour = genotype)) +
  geom_rect(aes(xmin = gDNA_min_mut_reads, xmax =Inf,  ymin = gDNA_control_VAF$MUT_threshold, ymax = Inf),
            fill = "#f0f0f0", colour = NA)+geom_point(position = "jitter")+
  geom_vline(aes(xintercept = gDNA_min_mut_reads),  linetype = "longdash")+
  geom_hline(aes(yintercept = gDNA_control_VAF$MUT_threshold),  linetype = "longdash") +
  geom_hline(aes(yintercept = gDNA_control_VAF$WT_threshold),  linetype = "longdash") +
  labs(title = paste(mutation_name, "gDNA threshold"), y = "VAF", x = "Mutant reads")+
  scale_x_log10()+scale_y_log10()


ggplot(genotyping_mRNA_filtered, aes(x = !!mut+0.1, y = VAF+0.0001, colour = Sample)) +
  geom_rect(aes(xmin = mRNA_min_mut_reads, xmax =Inf,  ymin = mRNA_control_VAF$MUT_threshold, ymax = Inf),
            fill = "#f0f0f0", colour = NA)+geom_point(position = "jitter")+
  geom_vline(aes(xintercept = mRNA_min_mut_reads),  linetype = "longdash")+
  geom_hline(aes(yintercept = mRNA_control_VAF$MUT_threshold),  linetype = "longdash") +
  geom_hline(aes(yintercept = mRNA_control_VAF$WT_threshold),  linetype = "longdash") +
  labs(title = paste(mutation_name, "mRNA threshold"), y = "VAF", x = "Mutant reads")+
  scale_x_log10()+scale_y_log10()
ggplot(genotyping_mRNA_filtered %>% filter(Sample !=control), 
       aes(x = !!mut+0.1, y = VAF+0.0001, colour = genotype)) +
  geom_rect(aes(xmin = mRNA_min_mut_reads, xmax =Inf,  ymin = mRNA_control_VAF$MUT_threshold, ymax = Inf),
            fill = "#f0f0f0", colour = NA)+geom_point(position = "jitter")+
  geom_vline(aes(xintercept = mRNA_min_mut_reads),  linetype = "longdash")+
  geom_hline(aes(yintercept = mRNA_control_VAF$MUT_threshold),  linetype = "longdash") +
  geom_hline(aes(yintercept = mRNA_control_VAF$WT_threshold),  linetype = "longdash") +
  labs(title = paste(mutation_name, "mRNA threshold"), y = "VAF", x = "Mutant reads")+
  scale_x_log10()+scale_y_log10()

# Print the number of mutant and WT cells 
genotyping_gDNA_filtered %>% 
  group_by(Sample, genotype) %>% 
  count() %>% 
  spread(genotype, n)

genotyping_mRNA_filtered %>% 
  group_by(Sample, genotype) %>% 
  count() %>% 
  spread(genotype, n)




##### CONSENSUS between gDNA and mRNA amplicons ######

# Join the mutation calling to the full table
genotyping_gDNA <- genotyping_gDNA_filtered %>% 
  select(Cell, genotype) %>% 
  right_join(genotyping_gDNA)

genotyping_mRNA <- genotyping_mRNA_filtered %>% 
  select(Cell, genotype) %>% 
  right_join(genotyping_mRNA)

# Fill in the undetected cells as Undetected
genotyping_gDNA <- genotyping_gDNA %>% 
  mutate(genotype = replace_na(genotype,"Undetected"))
genotyping_mRNA <- genotyping_mRNA %>% 
  mutate(genotype = replace_na(genotype,"Undetected"))

# Create a summary table
genotyping_gDNA_summary <- genotyping_gDNA %>% 
  select(Cell, Sample, Sample_type, Plate, mutation, coverage, !!ref, !!mut, VAF, genotype) %>% 
  rename(Ref = !!ref, 
         Mut = !!mut)
genotyping_mRNA_summary <- genotyping_mRNA %>% 
  select(Cell, Sample, Plate, mutation, coverage, !!ref, !!mut, VAF, genotype) %>% 
  rename(Ref = !!ref, 
         Mut = !!mut)

# Join the gDNA and mRNA tables together
genotyping_joined <- genotyping_gDNA_summary %>% 
  full_join(genotyping_mRNA_summary, by = c("Cell", "Sample", "Plate", "mutation"), suffix = c(".gDNA", ".mRNA"))


# Call a consensus genotype based on gDNA and mRNA
# 1. If the mutation was identified in either the gDNA or cDNA amplicon, the cell is called ‘mutant’.
# 2. If both amplicons were WT, the cell is called ‘WT’.
# 3. If the gDNA amplicon was WT but the cDNA amplicon was undetected or undetermined, the cell is called ‘WT’.
# 4. If the cDNA amplicon was WT but the gDNA amplicon was undetected or undetermined, the cell is called ‘undetermined’,
# due to the high allelic dropout rate of cDNA amplicons.


genotyping_joined <- genotyping_joined %>% 
  mutate(genotype = case_when(
    # If gDNA is MUT, cell is MUT
    genotype.gDNA == "Mut" & genotype.mRNA %in% c("Undetected","Undetermined", "WT", "Mut") ~ "Mut",
    # If mRNA is MUT, cell is MUT
    genotype.gDNA %in% c("Undetected","Undetermined", "WT", "Mut") & genotype.mRNA == "Mut" ~ "Mut",
    # If both are WT, cell is WT
    genotype.gDNA == "WT" & genotype.mRNA == "WT" ~ "WT",
    # IF gDNA is WT and mRNA anything but Mut, cell is WT
    genotype.gDNA == "WT" & genotype.mRNA %in% c("WT", "Undetected", "Undetermined") ~ "WT",
    # If gDNA is undetected or undetermined, we call cell undetermined
    genotype.gDNA == "Undetected" & genotype.mRNA %in% c("Undetected","Undetermined", "WT") ~ "Undetected",
    genotype.gDNA == "Undetermined" & genotype.mRNA %in% c("WT", "Undetected", "Undetermined") ~ "Undetermined",
    TRUE ~ "NA"))


## Write genotyping calls to a summary file ####

genotyping_joined %>% write_tsv(paste("data/genotyping/sample_", sample, "/genotyping_summary_", sample, "_", mutation_name, ".tsv", sep = "")) 


genotyping_submitted <- read_tsv("~/OneDrive - Nexus365/CHIP Project/Paper/Data submission/figshare/TARGET-seq_genotyping_data.tsv")

mismatch <- genotyping_joined %>% 
  left_join(select(genotyping_submitted, Cell, mutation, genotype) %>% rename(genotype.paper = genotype) %>% filter(mutation == mutation_name)) %>% 
  filter(genotype != genotype.paper)

